grabbit/src/main/groovy/com/twcable/grabbit/jcr/ProtoPropertyDecorator.groovy

```diff
@@ -0,0 +1,92 @@
+import com.twcable.grabbit.proto.NodeProtos.Property as ProtoProperty
+import com.twcable.grabbit.proto.NodeProtos.Value as ProtoValue
+import groovy.transform.CompileStatic
+import org.apache.jackrabbit.value.ValueFactoryImpl
+
+import javax.annotation.Nonnull
+import javax.jcr.Node as JCRNode
+import javax.jcr.Property
+import javax.jcr.PropertyType
+import javax.jcr.Value
+import javax.jcr.ValueFormatException
+
+import static org.apache.jackrabbit.JcrConstants.JCR_MIXINTYPES
+import static org.apache.jackrabbit.JcrConstants.JCR_PRIMARYTYPE
+
+@CompileStatic
+class ProtoPropertyDecorator {
+
+    @Delegate
+    ProtoProperty innerProtoProperty
+
+    ProtoPropertyDecorator(@Nonnull ProtoProperty protoProperty) {
+        this.innerProtoProperty = protoProperty
+    }
+
+
+    void writeToNode(@Nonnull JCRNode node) {
+        if(primaryType || mixinType) throw new IllegalStateException("Refuse to write jcr:primaryType or jcr:mixinType as normal properties.  These are not allowed")
+        try {
+            if (!innerProtoProperty.hasValues()) {
+                node.setProperty(innerProtoProperty.name, getPropertyValue(), innerProtoProperty.type)
+            }
+            else {
+                Value[] values = getPropertyValues()
+                node.setProperty(innerProtoProperty.name, values, innerProtoProperty.type)
+            }
+        }
+        catch (ValueFormatException ex) {
+            if (ex.message.contains("Multivalued property can not be set to a single value")) {
+                //If this is the exception, that means that a property with the name already exists
+                final Property currentProperty = node.getProperty(innerProtoProperty.name)
+                if (currentProperty.multiple) {
+                    final Value[] values = [getPropertyValue()]
+                    node.setProperty(innerProtoProperty.name, values, innerProtoProperty.type)
+                }
+            }
+            else if (ex.message.contains("Single-valued property can not be set to an array of values")) {
+                node.setProperty(innerProtoProperty.name, getPropertyValues().first(), innerProtoProperty.type)
+            }
+        }
+    }
+
+
+    boolean isPrimaryType() {
+        innerProtoProperty.name == JCR_PRIMARYTYPE
+    }
+
+
+    boolean isMixinType() {
+        innerProtoProperty.name == JCR_MIXINTYPES
+    }
+
+
+    private Value getPropertyValue() throws ValueFormatException {
+        getJCRValueFromProtoValue(innerProtoProperty.getValue())
+    }
+
+
+    private Value[] getPropertyValues() throws ValueFormatException {
+        return innerProtoProperty.values.valueList.collect { ProtoValue protoValue -> getJCRValueFromProtoValue(protoValue) } as Value[]
+    }
+
+
+    private Value getJCRValueFromProtoValue(ProtoValue value) throws ValueFormatException {
+
+        final valueFactory = ValueFactoryImpl.getInstance()
+
+        if (innerProtoProperty.type == PropertyType.BINARY) {
```

Comment:
In this specific situation, the comparison of a value to an enumerated list of potential values is conventionally handled with a `switch`, so I lean a little toward that here. If it were a single `if else` it would be easier/clearer to stay with the `if` structure. However, since this is `if else {if else}` it's inching into table-lookup territory.